### DISCUSSION TITLE
Totals in table charts

### DISCUSSION THREAD

**Discussion by michael-s-molina** [apache/superset#29297] [General]
Our users observed some problems with Table chart totals and I would like to get input on how to resolve them.

Consider this chart where we have a dimension and one metric that uses `DISTINCT`:

<img width="1470" alt="Screenshot 2024-06-18 at 15 38 33" src="https://github.com/apache/superset/assets/70410625/2bd89a29-23f3-41d5-a0e5-47af790583d0">

We can see that the `COUNT_DISTINCT(state)` total does not match the sum of the rows. 

This happens because the total is currently being calculated independently of the results of the first query:

<img width="870" alt="Screenshot 2024-06-18 at 15 40 14" src="https://github.com/apache/superset/assets/70410625/fcec4e1c-a03d-46b5-8264-0c315fbb6315">

The second query refers to the total and completely ignores the group by on the first query which is problematic when the metrics use some SQL constructs such as `DISTINCT`, `MAX`, `AVG`, etc.

Another problematic example is when Row Limit is applied:

<img width="1471" alt="Screenshot 2024-06-18 at 15 45 04" src="https://github.com/apache/superset/assets/70410625/a4d7b82f-0b6b-4004-8e17-081ee4646c66">

In this example, I applied a row limit of 2 and neither total represents the sum of the column values. Although the `Show totals` control has the following tooltip, it still provides a poor UX for users:

> Show total aggregations of selected metrics. Note that row limit does not apply to the result.

Another example is when we have a paginated table, where the totals are also not matching the sum of the column values:

<img width="1467" alt="Screenshot 2024-06-18 at 15 49 48" src="https://github.com/apache/superset/assets/70410625/0cfa6b91-8540-45dc-a2ac-ac7a136065af">

<br>Some possible solutions:

1 - Calculate the totals with only what's displayed. This means that the total would vary depending on the row limit, dimensions, SQL constructs and selected page. The downside is that we wouldn't have the total of all data but that could be easily achieved with another chart.

2 - Calculate the totals on top of the results of the first query instead of an independent query. We would still need to decide what to do with the row limit. Should the totals be calculated considering the row limit?

3 - Make all this configurable. This would increase code complexity and may also decrease UX because the total calculation will differ between charts. We would need a way to inform users of the configuration when displaying data.

Thoughts @rusackas @yousoph @kgabryje @eschutho @mistercrunch?

**Comment by rusackas** [comment]
I vote for #1 - it seems like the simpler approach (#2 adds complexity with effectively the same result, if I understand corerctly). I believe that the totals should not reflect the dataset, but rather the table as it's displayed (with filters, limits, grouping, etc. applied).

**Comment by john-bodley** [comment]
@michael-s-molina per your comment,

> We can see that the COUNT_DISTINCT(state) total does not match the sum of the rows.

which is what one would expect for said metric. Maybe the term `Totals` should be renamed to `Aggregates` (or similar) and include a footnote (if necessary).

**Comment by mistercrunch** [comment]
There's two challenges here:
- non-additive metrics (like count distinct, but also applies to things averages and many other agg functions)
- partial counts / limits where the client has a limited view on total

In both cases, computation should be done database-side as either cases could blow up the backend/frontend. 

Also in both cases, totals of this kind forces 2-phases or subquery. 

The existing feature "series limit" where one can say "I want to see the full time series for the top N {countries}" does it with a subquery to first identify the top N countries and then fetch their respective time series. I remember building the feature and though I could add a checkbox to "Also lump/aggregate others members" where it would show the top N times series + another time series for "all other countries"

Similarly, our LIMIT feature could offer something similar where it could lump the rest of the rows without its dimensionality, making computing a total possible. But this doesn't help with non-additive metrics.

For non-additive metrics, you probably want to `UNION ALL` a dimension-free total query while applying the same WHERE clause. For a lot of database that will tend to be two table scans. Also, how do we know whether an metric is non-additive? I think it's easy for the ones we generate, but for the user-provided SQL expressions, we'd probably need a "treat as a non-additive metric".

**Comment by michael-s-molina** [comment]
> For non-additive metrics, you probably want to UNION ALL a dimension-free total query while applying the same WHERE clause.

@mistercrunch This is the current behavior but the problem is that only the `WHERE` clause is not sufficient because non-additive metrics are influenced by the `GROUP BY` too. Check the first screenshot.

**Comment by michael-s-molina** [comment]
Looking at the comments, we can see that we don't have a clear alignment. Let's go back a little bit and talk about our expectations (requirements):

1 - Take the screenshot 1 as an example where the `COUNT_DISTINCT(state)` is influenced by the `GROUP BY`. What do you expect the totals to be? The sum of all rows or the result that is being showed (16)?

2- Take the screenshot 3 as an example where there's a `LIMIT` of 2 applied. Do you expect the totals to reflect the limit you imposed or do you want to see the total of all rows independently of the `LIMIT`?

**Reply by rusackas** [reply]
Perhaps I'm naïve, but in the case of a table, I'd expect the total to be the sum of all the rows displayed above it. Same goes for any other aggregation. I see it as an aggregation of the table being displayed, not some alternate query of the underlying database table.

**Comment by michael-s-molina** [comment]
We agreed in a meeting to replace the Total nomenclature with Summary and add a tooltip in the table with more details. Thanks for the great discussion @villebro @john-bodley @mistercrunch @rusackas. I'll open a PR with the results of our discussion.

**Comment by rea725** [comment]
I have the same type of circumstance here.  For our situation, I think solution 2 proposed by @michael-s-molina would be best.  I am guessing this is the approach taken by other tools such as Tableau, PowerBI, and even Excel pivot tables.  That's just a guess though, not confirmed.  

I think this might be a case where staying in sync with the way other tools do this could be the best approach.  Option 3 seems like it'd be a lot of extra work on the front end, and would still potentially lead to UX confusion.  I'm not dead-set against it, though.

<img width="1322" height="315" alt="image" src="https://github.com/user-attachments/assets/b02fed79-07ea-41ce-a51d-d7820a34cafb" />

**Comment by rusackas** [comment]
Bit of triage on the totals saga: I've consolidated the scattered bug reports (#38674, #32260, #37627, #34350/#34425/#34426...) onto #25747 as the canonical tracking issue, scoped to both pivot and table charts. This thread stays the design home though, since the open question is still multi-query-per-level (#34592's approach, with perf tradeoffs) vs. a formula-aware/non-additive-flag approach. CC @michael-s-molina since you wrote up the original root-cause here.

**Comment by rusackas** [comment]
Circling back here: the totals saga now has a formal writeup at #41463, with a working POC in #41184 (DB-computed totals per grouping level, `GROUPING SETS` where supported, fallback otherwise). This thread stays the design backstory, but further discussion can move over to the SIP. CC @michael-s-molina since you kicked this off.
